src/jenkins.py

```diff
@@ -705,3 +709,195 @@ def get_agent_name(unit_name: str) -> str:
+    dependency_lookup: dict[str, tuple[str, ...]] = defaultdict(lambda: ())  # pragma: no cover
+    for line in plugin_dependency_outputs:
+        match = re.match(PLUGIN_LINE_CAPTURE, line)
+        if not match:
+            continue
+        plugin, dependencies = match.group(1), match.group(3)
+        if not dependencies:
+            dependency_lookup[plugin] = ()
+        try:
+            dependency_lookup[plugin] = tuple(
+                _get_plugin_name(dependency) for dependency in dependencies.split(", ")
+            )
+        except ValidationError as exc:
+            logger.error("Invalid plugin dependency, %s", exc)
+            continue
+
+    return dependency_lookup
+
+
+def _set_jenkins_system_message(message: str, container: ops.Container) -> None:
+    """Set a system message on Jenkins.
+
+    Args:
+        message: The system message to display.
+        container: The Jenkins workload container.
+    """
+    jcasc_yaml = container.pull(JCASC_CONFIG_FILE_PATH, encoding="utf-8").read()
+    config = yaml.safe_load(jcasc_yaml)
+    config["jenkins"]["systemMessage"] = message
+    container.push(
+        JCASC_CONFIG_FILE_PATH, yaml.dump(config), encoding="utf-8", user=USER, group=GROUP
+    )
+
+
+def _traverse_dependencies(
+    plugin: str, dependency_lookup: typing.Mapping[str, typing.Iterable[str]], seen: set[str]
+) -> typing.Iterable[str]:
+    """Recursively traverse through plugin dependencies.
+
+    Args:
+        plugin: The plugin to recurse through the dependencies.
+        dependency_lookup: The plugin and its dependencies lookup table.
+        seen: a set indicating whether a plugin has already been visited.
+
+    Yields:
+        Plugin and its dependents.
+    """
+    if plugin in seen or plugin not in dependency_lookup:
+        return
+    yield plugin
+    seen.add(plugin)
+    for dependency in dependency_lookup[plugin]:
+        yield from _traverse_dependencies(dependency, dependency_lookup, seen)
+
+
+def _get_allowed_plugins(
+    top_level_plugins: typing.Iterable[str],
+    dependency_lookup: typing.Mapping[str, typing.Iterable[str]],
+) -> typing.Iterable[str]:
+    """Get the plugin short names of allowed plugins and their dependencies.
+
+    Args:
+        top_level_plugins: The allowed plugins short names to add to allowed plugins with their
+            dependencies.
+        dependency_lookup: The plugin dependency lookup table.
+
+    Yields:
+        The allowed plugin short name.
+    """
+    seen: set[str] = set()
+    for plugin in top_level_plugins:
+        if plugin not in seen:
+            yield from _traverse_dependencies(plugin, dependency_lookup, seen)
+
+
+def _get_top_level_plugins(
+    plugins: typing.Iterable[str], dependency_lookup: typing.Mapping[str, typing.Iterable[str]]
+) -> typing.Iterable[str]:
+    """Get top level plugins that are not dependencies of other plugins.
+
+    Args:
+        plugins: Plugins to extract top level plugins from.
```

Comment:
Why is this argument needed? Can't this be built entirely based on `dependency_lookup`?

Reply:
This is a list of plugins that represents a set all plugins (plugins + dependencies) and it is used to find out which plugin is a top level plugin `(plugin + dependencies) - dependencies` that should be present when showing Jenkins system message. I'm not sure I could come up with a better name for this variable. Perhaps `plugin_and_dependencies`?

Reply:
Renamed the function. Thank you :)